File: js/sdk-core/src/tx-state.ts
import { field, array, variant, struct, option } from '@race-foundation/borsh'

export abstract class TxState {}

export type TxStateKind = 'PlayerConfirming' | 'PlayerConfirmingFailed' | 'SettleSucceed'

export interface ITxStateKind {
    kind(): TxStateKind
}

export class ConfirmingPlayer {
    @field('u64')
    id!: bigint
    @field('string')
    addr!: string
    @field('u16')
    position!: number
    @field('u64')
    balance!: bigint

    constructor(fields: any) {
        Object.assign(this, fields)
    }
}

@variant(0)
export class PlayerConfirming extends TxState implements ITxStateKind {
    @field(array(struct(ConfirmingPlayer)))
    confirmPlayers!: ConfirmingPlayer[]
    @field('u64')
    accessVersion!: bigint

    constructor(fields: any) {
        super()
        Object.assign(this, fields)
    }
    kind(): TxStateKind {
        return 'PlayerConfirming'
    }
}

@variant(1)
export class PlayerConfirmingFailed extends TxState implements ITxStateKind {
    @field('u64')
    accessVersion!: bigint

    constructor(fields: any) {
        super()
        Object.assign(this, fields)
    }
    kind(): TxStateKind {
        return 'PlayerConfirmingFailed'
    }
}

@variant(2)
export class SettleSucceed extends TxState implements ITxStateKind {
    @field('u64')
    settleVersion!: bigint
    @field(option('string'))
    signature: string | undefined

    constructor(fields: any) {
        super()
        Object.assign(this, fields)
    }
    kind(): TxStateKind {
        return 'SettleSucceed'
    }
}

File: core/src/types/tx_state.rs
use crate::types::PlayerJoin;
use borsh::{BorshDeserialize, BorshSerialize};
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, BorshDeserialize, BorshSerialize)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub struct ConfirmingPlayer {
    pub id: u64,
    pub addr: String,
    pub position: u16,
}

impl From<PlayerJoin> for ConfirmingPlayer {
    fn from(value: PlayerJoin) -> Self {
        Self {
            id: value.access_version,
            addr: value.addr,
            position: value.position,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, BorshDeserialize, BorshSerialize)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub enum TxState {
    PlayerConfirming {
        confirm_players: Vec<ConfirmingPlayer>,
        access_version: u64,
    },

    PlayerConfirmingFailed(u64),

    SettleSucceed {
        settle_version: u64,
        signature: Option<String>,
    },
}
